File: go/166/main.go
package main

import (
	"fmt"
)

func findMax(nums []int) int {
    maxVal := nums[0]
    for _, num := range nums {
        if num > maxVal {
            maxVal = num
        }
    }
    return maxVal
}

func main() {
    fmt.Println(findMax([]int{3, 1, 4, 1, 5, 9, 2}))   // 9
    fmt.Println(findMax([]int{-7, -3, -10, -2}))       // -2
    fmt.Println(findMax([]int{42}))                    // 42
}

File: python/166/main.py
from typing import List


class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        if not nums:
            return 0

        max_prod = nums[0]
        min_prod = nums[0]
        result = nums[0]

        for i in range(1, len(nums)):
            current = nums[i]
            temp_max = max(current, max_prod * current, min_prod * current)
            min_prod = min(current, max_prod * current, min_prod * current)
            max_prod = temp_max
            result = max(result, max_prod)

        return result


# テスト
sol = Solution()
print(sol.maxProduct([2, 3, -2, 4]))
print(sol.maxProduct([-2, 0, -1]))
print(sol.maxProduct([-2, 3, -4]))
